text/0000-more-api-documentation-conventions.md

```diff
@@ -0,0 +1,409 @@
+
+## Form
+[form]: #form
+
+These conventions relate to the formatting of the documentation, how they
+appear in source code.
+
+### Use line comments
+[use-line-comments]: #use-line-comments
+
+Avoid block comments. Use line comments instead:
+
+```rust
+// Wait for the main task to return, and set the process error code
+// appropriately.
+```
+
+Instead of:
+
+```rust
+/*
+ * Wait for the main task to return, and set the process error code
+ * appropriately.
+ */
+```
+
+Only use inner doc comments `//!` to write crate and module-level documentation,
+nothing else. When using `mod` blocks, prefer `///` outside of the block:
+
+```rust
+/// This module contains tests
+mod test {
+    // ...
+}
+```
+
+over
+
+```rust
+mod test {
+    //! This module contains tests
+
+    // ...
+}
+```
+
+### Using Markdown
+[using-markdown]: #using-markdown
+
+Within doc comments, use Markdown to format your documentation.
+
+Use top level headings # to indicate sections within your comment. Common headings:
+
+* Examples
+* Panics
+* Errors
+* Safety
+* Aborts
+* Undefined Behavior
+
+Even if you only include one example, use the plural form: ‘Examples’ rather
+than ‘Example’. Future tooling is easier this way.
+
+Use graves (`) to denote a code fragment within a sentence.
+
+Use triple graves (```) to write longer examples, like this:
+
+    This code does something cool.
+
+    ```rust
+    let x = foo();
+
+    x.bar();
+    ```
+
+When appropriate, make use of Rustdoc’s modifiers. Annotate triple grave blocks with
+the appropriate formatting directive.
+
+    ```rust
+    println!("Hello, world!");
+    ```
+
+    ```ruby
+    puts "Hello"
+    ```
+
+In API documentation, feel free to rely on the default being ‘rust’:
+
+    /// For example:
+    ///
+    /// ```
+    /// let x = 5;
+    /// ```
+
+In long-form documentation, always be explicit:
+
+    For example:
+
+    ```rust
+    let x = 5;
+    ```
+
+This will highlight syntax in places that do not default to ‘rust’, like GitHub.
+
+Rustdoc is able to test all Rust examples embedded inside of documentation, so
+it’s important to mark what is not Rust so your tests don’t fail.
+
+References and citation should be linked ‘reference style.’ Prefer
+
+```
+[Rust website]
+
+[Rust website]: http://www.rust-lang.org
+```
+
+to
+
+```
+[Rust website](http://www.rust-lang.org)
+```
+
+### Examples in API docs
+[examples-in-api-docs]: #examples-in-api-docs
+
+Everything should have examples. Here is an example of how to do examples:
+
+```
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, "bar");
+/// ```
+///
+/// Passing a closure to compare with, rather than a string:
+///
+/// ```
+/// use op;
+///
+/// let s = "foo";
+/// let answer = op::compare(s, |a| a.chars().is_whitespace().all());
+/// ```
+```
+
+For particularly simple APIs, still say “Examples” and “Basic usage:” for
+consistency’s sake.
```

Comment:
I agree with @tshepang that the "Basic usage" detracts more than it adds.

It is not currently used consistently in the standard library. And when I look at examples that include it, and examples that don't, I prefer the latter. For example, [`str::len()`] vs. [`Vec::len()`].

Now, if there is any prose discussion needed for an example, it makes sense to have some introductory text, or text describing each example in turn. But the string "Basic usage" just takes up vertical space, without adding much. The `# Examples` heading already sets this off as an example, and most examples can be fairly basic, so it seems like doing this just for the sake of consistency will waste a lot of space.

Even for those examples that do have some introductory text for more specific examples later, I don't think the "Basic usage" adds much. The [`str::chars()`] example would read just as well without it.

I decided to look to see if other style guides addressed examples in doc comments. Not many do; the [JavaDoc style guide] does not, nor does [Python's PEP-257]. The [Python `doctest` documentation] does not separate examples into a separate section, but intersperses them with the rest of the documentation. The [Ruby on Rails API Documentation Guide] recommends putting examples at the end, and makes an "Examples" heading optional.

I think that there are good arguments both ways for the "Examples" heading being optional or mandatory for consistency; I don't see a problem with requiring that. But requiring both an "Examples" heading and "Basic usage" seems like overkill.

[`str::len()`]: http://doc.rust-lang.org/std/primitive.str.html#method.len
[`Vec::len()`]: http://doc.rust-lang.org/collections/vec/struct.Vec.html#method.len
[`str::chars()`]: http://doc.rust-lang.org/std/primitive.str.html#method.chars
[JavaDoc style guide]: http://www.oracle.com/technetwork/articles/java/index-137868.html
[Python's PEP-257]: https://www.python.org/dev/peps/pep-0257/
[Python `doctest` documentation]: https://docs.python.org/2.7/library/doctest.html
[Ruby on Rails API Documentation Guide]: http://guides.rubyonrails.org/api_documentation_guidelines.html#example-code